src/Processors/Transforms/FilterSortedStreamByRange.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <Interpreters/ExpressionActions.h>
+#include <Processors/ISimpleTransform.h>
+#include <Processors/Transforms/ExpressionTransform.h>
+#include <Processors/Transforms/FilterTransform.h>
+
+namespace DB
+{
+
+/** Could be used when the predicate given by expression_ is true only on one continuous range of values and input is monotonous by that value.
+ *  The following optimization applies: when a new chunk of data comes in we firstly execute the expression_ only on the first and the last row.
+ *  If it evaluates to true on both rows then the whole chunk is immediately passed to further steps.
+ *  Otherwise, we apply the expression_ to all rows.
+ */
+class FilterSortedStreamByRange : public ISimpleTransform
+{
+public:
+    FilterSortedStreamByRange(
+        const Block & header_,
+        ExpressionActionsPtr expression_,
+        String filter_column_name_,
+        bool remove_filter_column_,
+        bool on_totals_ = false)
+        : ISimpleTransform(
+            header_,
+            FilterTransform::transformHeader(header_, expression_->getActionsDAG(), filter_column_name_, remove_filter_column_),
+            true)
+        , filter_transform(header_, expression_, filter_column_name_, remove_filter_column_, on_totals_)
+    {
+    }
+
+    String getName() const override { return "FilterSortedStreamByRange"; }
+
+    void transform(Chunk & chunk) override
+    {
+        int rows_before_filtration = chunk.getNumRows();
+        if (rows_before_filtration < 2)
+        {
+            filter_transform.transform(chunk);
+            return;
+        }
+
+        // Evaluate expression on just the first and the last row.
+        // If both of them satisfies conditions, than skip calculation for all the rows in between.
```

Comment:
It is a little bit confusing to me.
Shouldn't we check first and last *granule* we read?

I would expect that this transform has:
* `begin_offset` (which is an offset for the first granule, ~8192)
* `end_offset` (which is is the start of the last granule, ~ (num rows in part) - 8192
* `num_processed_rows` counter

and we check that `if (num_processed_rows >= begin_offset && num_processed_rows + chunk.getNumRows() <= end_offset)` then we can skip filtration of a chunk.

Reply:
the idea was to make it reasonably reusable the task of filtering ordered streams by range looks fairly common.